Add new HsvColor struct

robloo opened the pull request:
## What does the pull request do?

Add a new HsvColor struct

Closes https://github.com/AvaloniaUI/Avalonia/issues/7705

## What is the current behavior?

No Hsv color model support.

## What is the updated/expected behavior with this PR?

Hsv color model support with HsvColor including conversion to Color and the RGB color model.

## How was the solution implemented (if it's not obvious)?

## Checklist

- [ ] Added unit tests (if possible)?
- [x] Added XML documentation to any related classes?
- [ ] Consider submitting a PR to https://github.com/AvaloniaUI/Documentation with user documentation

## Breaking changes

None

## Obsoletions / Deprecations

None

## Fixed issues

Closes https://github.com/AvaloniaUI/Avalonia/issues/7705

robloo committed:
Add new HsvColor struct

robloo committed:
Add Empty support in Color

robloo commented on src/Avalonia.Visuals/Media/Color.cs:
```diff
@@ -14,6 +14,11 @@ namespace Avalonia.Media
+        /// <summary>
+        /// Represents a null <see cref="Color"/> with zero set for all channels.
+        /// </summary>
+        public static readonly Color Empty = new Color();
```
I was surprised this didn't already exist in Avalonia. It is the same as WPF now.

https://docs.microsoft.com/en-us/dotnet/api/system.drawing.color.empty?view=net-6.0

robloo replied:
Marking resolved since no additional comments in review

MarchingCube replied:
I was actually surprised that this is a thing in WPF, since from my perspective "empty" color makes no sense semantically and I wonder how this is intended to be used. But then I checked WPF color and it has no such weird members https://docs.microsoft.com/en-us/dotnet/api/system.windows.media.color?view=windowsdesktop-6.0.

System.Drawing docs about "Represents a color that is null." are a total gibberish to me and looks like a poor man's nullable color.

I vote against having this (and `IsEmpty`) in the public API since it is just confusing. @AvaloniaUI/core

robloo replied:
I've needed this in the past. A good example is a color picker where it's necessary to know if the user made a selection or not. This also exists in the System.Drawing.Color class.

robloo replied:
`Nullable<HsvColor>` is similar as you pointed out; however, there is no need to have such overhead when the color struct itself can represent such a case. It cleans up a lot of API's that would otherwise need a nullable color to communicate such information.

robloo replied:
Just did a search for "Color.IsEmpty" online as well. A LOT of implementations do this as well.

MarchingCube replied:
I mostly checked modern frameworks and didn't find it anywhere:
- https://api.flutter.dev/flutter/dart-ui/Color-class.html
- https://developer.apple.com/documentation/swiftui/color
- https://doc.qt.io/qt-5/qcolor.html

Do you have an example from other framework? I would be curious to see their reasoning.

robloo replied:
This is still current in .NET. I didn't check other frameworks.

https://github.com/dotnet/runtime/blob/main/src/libraries/System.Drawing.Primitives/src/System/Drawing/Color.cs

Note that `IsEmpty`; however, is tracked with a separate "state" field. That would mitigate your concern about it being equal to #00000000 which is technically a valid color.

My understanding was in older .NET Framework it was implemented as I did here. However, that may be wrong:
https://referencesource.microsoft.com/#System.Drawing/commonui/System/Drawing/Color.cs,1433

Here are some .NET examples:

https://docs.microsoft.com/en-us/dotnet/api/system.drawing.color.empty?view=net-6.0
https://help.syncfusion.com/cr/file-formats/Syncfusion.Presentation.IColor.html#Syncfusion_Presentation_IColor_IsEmpty
https://www.grapecity.com/activereportsnet/docs/v14/online/GrapeCity.ActiveReports.Viewer.Wpf~GrapeCity.Viewer.Common.Color~Empty.html

It was even done for other types:

https://docs.microsoft.com/en-us/dotnet/api/system.drawing.point.isempty?view=net-6.0
https://docs.microsoft.com/en-us/dotnet/api/system.drawing.rectangle.empty?view=net-6.0

maxkatz6 replied:
If you need more opinions, I also don't see much sense in Empty/IsEmpty here. 
For "#00000000" state we already have Transparent value.

It would make more meaning if we had some special flag as in winforms, but we don't. Also I am not sure if we should have it, might it affect interoperability? In any way special flag discussion (as well as adding IsEmpty after that) is out of this PR imo.

robloo replied:
> For "#00000000" state we already have Transparent value

These are not the same as mentioned above. #00000000 != #00FFFFFF. Transparent was intentionally set to this value for this reason.

Aside from the misunderstanding with the field, I certainly think this is still useful. You guys are really hard to convince sometimes though :) I'm not inventing anything new here.

maxkatz6 replied:
Moreover, do we **really** need that kind of information in Color struct? Where can it be used, where Transparent or nullable value cannot be used?

maxkatz6 replied:
> #00000000 != #00FFFFFF

That's another confusing moment. Understanding of what is transparent isn't consistent across frameworks. General understanding is both (or any color with 0 alpha channel).

In wpf it's #00ffffff, when in cocoa it's called "clear"  and has value of #00000000, and in html5 there is "transparent" with value of "rgba(0,0,0,0)".

robloo replied:
> Understanding of what is transparent isn't consistent across frameworks.

Point taken. But transparency isn't really what this is about.

All platforms should render #00FFFFFF as transparent too. Obviously, you don't translate `KnownColors.Transparent` for each backend. You have it defined once with #00FFFFFF as in .NET and it works well enough.

Concerns about different interpretations of empty is solvable by adding the `int state` field to track initialization and empty status. Standardizing what an Empty color means across platforms might actually be an argument in favor of keeping this functionality.

jmacato replied:
Sorry but i'll have to agree to the objections that @MarchingCube @maxkatz6 on the same grounds as well.

robloo replied:
Alright, that is 3 against 1 and no one other than myself is defending this -- which surprises me a little because again this isn't a new idea and there is past precedent. However, since it isn't required functionality right now I will remove it from HsvColor and Color and then create a separate issue to discuss and perhaps get wider attention and feedback over time.

I would prefer not to have a nullable ColorPicker in the future when this could all be managed in the Color structs themselves. I know performance and GDI+ reasons were a consideration when this was originally done as well.

Gillibald replied:
I can live with `Color.IsDefault`

robloo replied:
Color.IsDefault is not a standardized API... but it would convey the same purpose.

robloo replied:
I created a new issue: https://github.com/AvaloniaUI/Avalonia/issues/7762.

This is resolved but I'm leaving it unresolved so the links to this discussion work in the future.

robloo commented on src/Avalonia.Visuals/Media/Color.cs:
```diff
@@ -41,6 +46,18 @@ static Color()
+        /// Gets a value indicating whether this <see cref="Color"/> is uninitialized and
+        /// considered null.
+        /// </summary>
+        public bool IsEmpty
```
Again, from WPF and needed in certain advanced scenarios to differentiate an uninitialized color from Transparent (note that transparent is #00FFFFFF instead of #00000000.

https://docs.microsoft.com/en-us/dotnet/api/system.drawing.color.isempty?view=net-6.0

robloo replied:
Marking resolved since no additional comments in review

robloo commented on src/Avalonia.Visuals/Media/HsvColor.cs:
```diff
@@ -0,0 +1,493 @@
+        /// <param name="value">The value channel value in the HSV color model in the range from 0..1.</param>
+        /// <param name="alpha">The alpha channel value in the range from 0..1.</param>
+        /// <returns>A new RGB <see cref="Color"/> equivalent to the given HSVA values.</returns>
+        public static Color ToRgb(
```
Color model conversion code is well tested in the C# ports of the WinUI color picker and then again with this HsvColor struct on an internal project. This code is working fine in a real ColorPicker.

robloo replied:
Marking resolved since no additional comments in review

robloo commented:
Looks like build fails because of Clamp() and HashCode.Combine() usage. I thought Avalonia was on .net core 3.x minimum? Not sure I understand this.

I can't build Avalonia locally at the moment. It seems the project configurations have been broken for .NET Core over the past few weeks.

jmacato commented on src/Avalonia.Visuals/Media/HsvColor.cs:
```diff
@@ -0,0 +1,493 @@
+        /// </summary>
+        public static readonly HsvColor Empty = new HsvColor();
+
+        ////////////////////////////////////////////////////////////////////////
```
please avoid using this style of comments. use `/* */` instead.

jmacato commented on src/Avalonia.Visuals/Media/HsvColor.cs:
```diff
@@ -0,0 +1,493 @@
+        /// </summary>
+        public static readonly HsvColor Empty = new HsvColor();
+
+        ////////////////////////////////////////////////////////////////////////
```
please avoid using these style of comments. use /* */ instead.

robloo replied:
Has that been documented and agreed for the Avalonia code-base? Is there a format guide?

The issue with /* */ comments is that they block larger /* */ comments. If you ever need to comment-out code at least temporarily, /* */ quickly gets in the way. For that reason, the internal convention is to use // for all comments that should remain in code and /* */ is reserved for code-blocks that are temporarily disabled.

jmacato replied:
It's just convention. I just havent seen this kind of comment style being committed in all my years here.

robloo replied:
Yea, section headings aren't widely used. It is extremely useful to keep code organized so each dev doesn't just puts code wherever they like when they modify the file. Putting similar code together improves readability and maintainability quite a bit. I think this is used more in the embedded world where the requirements are far more strict and things like AUTOSAR and detailed code checking, design reviews and proofs mean perfect code is topmost priority. Over the air updates didn't always exist... I digress.

Since HsvColor will likely get no major changes in its life, I can remove these headings if you really don't like them. It would be better than using `/* */` which would also be strange for you but have the added concern of making it difficult to disable code in the future.

robloo replied:
Well, #regions are different than this. They are frowned-upon these days for other reasons.

If everyone is against these regions I will remove them. I'm surprised no one wants to enforce the structure of code within objects a bit more though -- it is standard in some projects.

robloo replied:
Removed due to complaints :)

jmacato commented on src/Avalonia.Visuals/Media/HsvColor.cs:
```diff
@@ -0,0 +1,493 @@
+                    (byte)Math.Round(min * 255));
+            }
+
+            // If the chroma is not zero, then we need to continue.  The first step is to figure out
```
this is better served with the multiline comment as well.

jmacato commented on src/Avalonia.Visuals/Media/HsvColor.cs:
```diff
@@ -0,0 +1,493 @@
+            // Alpha channel support was added
+
+            // Normalize RGBA channel values into the 0..1 range used by this algorithm
+            double r = red / 255.0;
```
perhaps `255d` instead of `255.0`? Idk, optional :)

robloo replied:
Just developer preference really. 255.0 is more readable IMO

jmacato commented on src/Avalonia.Visuals/Media/HsvColor.cs:
```diff
@@ -0,0 +1,493 @@
+            // then the chroma is maximized - this is a pure yellow, no gray of any kind.
+            // On the other hand, if we have RGB = (128, 128, 128), then the chroma being zero
+            // implies that this color is pure greyscale, with no actual hue to be found.
+            double chroma = max - min;
```
```suggestion
            var chroma = max - min;
```

robloo replied:
Same as above... this is pedantic and shouldn't be picked up by reviewers IMO.

jmacato commented on src/Avalonia.Visuals/Media/HsvColor.cs:
```diff
@@ -0,0 +1,493 @@
+            //
+            // From these facts, you can see that we can retrieve the chroma by simply multiplying the saturation and the value,
+            // and we can retrieve the minimum of the RGB channels by subtracting the chroma from the value.
+            double chroma = saturation * value;
+            double min = value - chroma;
```
```suggestion
            var chroma = saturation * value;
            var min = value - chroma;
```

robloo replied:
I strongly prefer not to overuse `var` when it isn't explicitly clear. I know you prefer to use `var` for everything. `double` makes the type clear here. Types are actually quite important in this sort of code. Other places use `byte`.

robloo replied:
Well, this isn't worth debating here as long as I can keep the type clear elsewhere. Since this uses defined variables directly it can be inferred easily when reading through the full algorithm.

jmacato commented on src/Avalonia.Visuals/Media/HsvColor.cs:
```diff
@@ -0,0 +1,493 @@
+            double value,
+            double alpha = 1.0)
+        {
+            // Note: Conversion code is originally based on the C++ in WinUI (licensed MIT)
```
Add licensing terms on top of the file. Kinda like what we did here: https://github.com/AvaloniaUI/Avalonia/blob/master/src/Avalonia.Controls/Grid.cs#L1-L4

robloo replied:
That isn't required for MIT code. However, if project convention I can do it. Remember this is a C++ -> C# port then has custom modifications so it isn't a direct copy/paste (though is very similar).

jmacato replied:
project convention. i wouldnt be saying this and my other reviews if it's not.

robloo replied:
Ok, fair enough and that is standard for the most part. I added a header to the file making it clear. I'm also keeping the comments here as well.

jmacato reviewed (approved):
concept looks good, just needed some consistency changes :) good job! :)

jmacato reviewed (changes requested):
ah sorry i misclicked lol

jmacato commented:
@robloo i think it's because we're using netstandard2.0 for the base libs

robloo committed:
Add additional license comments for color conversion code

robloo commented:
> @robloo i think it's because we're using netstandard2.0 for the base libs

Is this going to change? .NET Standard 2.0 is extremely old now and I thought 2.1 superseded it anyway?

I would prefer not having to implement my own Clamp method. I can copy/paste the hash algorithm from Color if needed.

maxkatz6 commented:
> Is this going to change

Not anytime soon. We still need it for Mono support and some exotic targets that it supports.

maxkatz6 commented:
> I would prefer not having to implement my own Clamp method

https://github.com/AvaloniaUI/Avalonia/blob/0749f9a34ff21562bda223a58799a61f62109fce/src/Avalonia.Base/Utilities/MathUtilities.cs#L206

robloo committed:
Use var per maintainer request

Co-authored-by: Jumar Macato <16554748+jmacato@users.noreply.github.com>

robloo committed:
Use var per maintainer request

Co-authored-by: Jumar Macato <16554748+jmacato@users.noreply.github.com>

robloo commented on src/Avalonia.Visuals/Media/Color.cs:
```diff
@@ -14,6 +14,11 @@ namespace Avalonia.Media
 #endif
     readonly struct Color : IEquatable<Color>
```
@jmacato Why is public conditional? Should HsvColor do the same? (I noticed this as I went to switch HsvColor to readonly which was missed at first).

maxkatz6 replied:
This type can be used in Avalonia.Build.Tasks project as included source file, but it won't be exported from this assembly.
So it's basically internal inside of Avalonia.Build.Tasks, but public inside of Avalonia.Visuals.

maxkatz6 replied:
Avalonia.Build.Tasks uses it for compile time parsing/

jmacato replied:
uhm... im not really sure myself either haha. i think it's for CI/Release tests.

robloo replied:
Ok, well, I'm leaving HsvColor `public` until told otherwise. I know you were trying to combine a bunch of projects too to simplify things.

robloo replied:
I ended up copy/pasting what was done for Color. After referencing HsvColor from Color this was mandatory due to the way projects are setup.

robloo committed:
Use MathUtilities.Clamp() for .NET Standard 2.0 support

robloo committed:
Use custom hash algo for .NET Standard 2.0 support

robloo committed:
Readonly struct

robloo commented:
Another question: How is implicit XAML type conversion supported? It will be useful in the future to be able to bind an HsvColor value to a SolidColorBrush directly in XAML. I'm not sure where that type conversion is implemented.

amwx commented:
> Another question: How is implicit XAML type conversion supported? It will be useful in the future to be able to bind an HsvColor value to a SolidColorBrush (Color) directly in XAML. I'm not sure where that type conversion is implemented.

Xaml compiler, from my experience, doesn't like type mismatches - even if they're implicitly convertible, at least not without a `TypeConverter`. For ex: My `Color2` class (which is a blended Color struct for various color spaces) doesn't bind to a property of `Color`.

robloo commented:
@amwx 

Yea, TypeConverter is how it was done in WPF I think. If this is too much work I won't bother. I just thought it could help integration in the future.

Its an interesting idea to have a UniversalColor. In the color picker I do have a common ColorChannel enum that just specified 1,2,3,alpha. That simplifies some calculation. I guess doing that here would cause maintainability and use confusion though. In the future it might be interesting to move all color model conversion into a UniversalColor though and have it available for advanced users. I think Color and HsvColor would still be needed for common use cases.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018955-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018955-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

jmacato commented:
@robloo the best time to break API's is before the v11 release. It means today :)

jmacato commented:
Also it'd be great to have animator/transition support for this as well.

robloo commented:
> @robloo the best time to break API's is before the v11 release. It means today :)

Well, this is only a part of the ColorPicker but is also generally useful. I don't plan on any breaking changes for v11. It would have been great to know your release schedule and I probably would have planned to get the entire color picker in for v11. I noticed in the project tracker you have several open items left though.

> Also it'd be great to have animator/transition support for this as well.

Not currently planned. It shouldn't be a breaking change though?

robloo committed:
Add Color.ToHsv() helper method

robloo committed:
Standardize comments

robloo committed:
Switch public/internal conditionally to match Color

robloo commented on src/Avalonia.Visuals/Media/Color.cs:
```diff
@@ -273,8 +290,15 @@ public uint ToUint32()
+        /// Returns the HSV color model equivalent of this RGB color.
         /// </summary>
+        /// <returns>The HSV equivalent color.</returns>
+        public HsvColor ToHsv()
```
I added this helper conversion method to Color. Now the build is broken. Why HsvColor is unavailable to Color in certain situations is beyond me. This seems like a poorly configured project? Color and HsvColor are now completely the same scope, namespace and project. This should just work.

robloo replied:
I had to manually add HsvColor to `src/Avalonia.Build.Tasks/Avalonia.Build.Tasks.csproj` to fix this. I'm not sure why it's setup that way, seems if build tasks needs these types it would just pull in the whole project or use globbing.

MarchingCube replied:
Since .net does not have static linking we have to copy files to the compiler to remove dependency on Avalonia projects. I'm not sure if globbing can work in this case since we are only pulling in a specific part of dependencies (in certain cases only a part of the file).

robloo replied:
Alright, thanks for the details. I think this could have been configured a bit better but I won't bother with it further here in this PR.

robloo committed:
Add HsvColor to Avalonia.Build.Tasks.csproj

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018963-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018963-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Add missing alpha clamping caught in review

robloo commented on src/Avalonia.Visuals/Media/HsvColor.cs:
```diff
@@ -0,0 +1,514 @@
+        /// <param name="value">The Value channel value in the range from 0..1.</param>
+        /// <param name="alpha">The Alpha (transparency) channel value in the range from 0..1.</param>
+        public HsvColor(
+            double hue,
+            double saturation,
+            double value,
+            double alpha)
```
The last thing that was caught in an internal review was the order of parameters here. In the Color constructor and FromAhsv, FromArgb, etc, Alpha channel is the first parameter. This is somewhat standard in .NET and follows with pre-multiplied alpha ordering as well.

The reason alpha was put at the end is due to the way Windows Community Toolkit methods work. This type was originally a drop-in replacement for some of those methods: https://github.com/CommunityToolkit/WindowsCommunityToolkit/blob/b39d70f6ea4d739a470768c103c8af2c5b1e02ef/Microsoft.Toolkit.Uwp/Helpers/ColorHelper.cs#L312

This is the same as encountered here in the `HsvColor.ToRgb()` and `HsvColor.FromRgb()` methods because alpha channel is optional and therefore must be at the end.

As this would be a pretty bad breaking change in the future (undetectable errors), we need to be clear on parameter ordering here. @jmacato @maxkatz6 What are your thoughts? Should alpha be first always except for methods where it is optional?

robloo replied:
Another reason to keep alpha at the end is because there are two constructors: one with alpha, one without. With alpha at the end it doesn't change the position of the HSV channels and therefore there would theoretically be fewer errors switching between constructors.

robloo replied:
@maxkatz6 @MarchingCube @jmacato Any preference with this?

jmacato replied:
use alpha, no need to follow UWP/WPF conventions if it doesnt make any sense in our "local" context.

jmacato replied:
follow the conventions of Avalonia's Color class since Hsv is just another form of it.

robloo replied:
Alright will put it at the front as flagged to match Color. This will standardize the `HsvColor(A,H,S,V)` and `Color(A,R,G,B)` constructors as well as follow the ordering in `Color.FromArgb()` and `Hsv.FromAhsv()`. 

Just realize this is now different than the ordering in the new To/FromRGB methods within HsvColor because the alpha is optional and must be last due to c# rules.

robloo replied:
Unless there are objections I'm planning to remove both the `Color(H,S,V)` constructor and the `HsvColor.FromHsv(H,S,V)` static method.

 1. As stated, there is concern that parameters ordering differences could cause issues if developers aren't careful. AHSV vs. HSV is easy to be off-by-one in channel ordering when moving from HSV to AHSV methods for example.
 2. There are no situations where AHSV parameter equivalents can't be used in place of HSV versions. Worse-case, the caller will need to provide 1.0 for the alpha value.
 3. Color doesn't have RBG-only constructors/methods. Everything takes an alpha channel value as well.

robloo replied:
I've removed the code as mentioned. I wish the core team was a little more vocal about API-related discussions like this. These things are important to discuss but I realize a bit boring/trivial.

Marking as resolved

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018965-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018965-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Optimizations

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018979-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018979-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

MarchingCube commented on src/Avalonia.Visuals/Media/HsvColor.cs:
```diff
@@ -0,0 +1,515 @@
+            }
+
+            // We similarly clamp saturation, value and alpha between 0 and 1.
+            saturation = saturation < 0.0 ? 0.0 : saturation;
```
Could use `MathUtilities.Clamp` three times to remove some boilerplate.

amwx replied:
Might make sense to decorate MathUtilities.Clamp with `[MethodImpl(MethodImplOptions.AggressiveInlining)]` as System.Math does

robloo replied:
MathUtilities.Clamp is still called 4 times in the constructor at the end of this method. I'm not sure I can remove this safely as HsvColor needs to have validation to ensure parameter ranges. However, there is more overhead in doing this that isn't an issue in other types like Color where the parameter type can never be out of range.

robloo replied:
As a further optimization I added an internal-only constructor that can bypass channel parameter value clamping. This means MathUtilties.Clamp is never called in this method. Again, performance concerns in future color spectrum calculations.

MarchingCube reviewed (changes requested):
Idea of having HSV colors is good since it makes animations/color generation easier. Just a few comments about the implementation.

robloo commented on src/Avalonia.Visuals/Media/HsvColor.cs:
```diff
@@ -0,0 +1,515 @@
+        /// <inheritdoc/>
+        public bool Equals(HsvColor other)
+        {
+            if (object.Equals(other.A, A) == false) { return false; }
```
Yea, good catch.

robloo replied:
@MarchingCube Was updated, please mark as resolved if you have no more concerns here.

robloo committed:
Optimization from review

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018993-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0018993-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

jmacato committed:
Merge branch 'master' into hsvcolor

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019002-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019002-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo committed:
Remove Empty field and IsEmpty property from Color/HsvColor

robloo committed:
Merge branch 'hsvcolor' of https://github.com/robloo/Avalonia into hsvcolor

robloo committed:
Remove section headings

robloo committed:
Make alpha the first parameter in the constructor matching Color

robloo committed:
Fix parameter ordering

robloo committed:
Remove IsNaN checks from constructor

In XAML NaN is used in some cases so I wanted to cover this case. However, it isn't worth the performance cost in a constructor like this. NaN, Infinity, etc. will cause exceptions.

robloo committed:
Add internal constructor that can skip parameter clamping

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019006-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019006-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo commented on src/Avalonia.Visuals/Media/HsvColor.cs:
```diff
@@ -0,0 +1,478 @@
+            double saturation,
+            double value)
+        {
+            A = MathUtilities.Clamp(alpha,      0.0, 1.0);
+            H = MathUtilities.Clamp(hue,        0.0, 360.0);
+            S = MathUtilities.Clamp(saturation, 0.0, 1.0);
+            V = MathUtilities.Clamp(value,      0.0, 1.0);
```
Marking as resolved since no one had additional comments here.

robloo replied:
Note: Maui is clamping too https://github.com/dotnet/Microsoft.Maui.Graphics/blob/cc1e48974dcad6b7c976cf1371fa73db23f677dd/src/Microsoft.Maui.Graphics/Color.cs#L30-L35

robloo committed:
Remove constructor and method with only HSV channels

Only variants with an alpha channel (AHSV) are supported. This is to avoid concerns about accidental parameter ordering and the API matches with Color.

robloo committed:
Simplify

robloo committed:
Merge branch 'master' into hsvcolor

robloo commented:
There are no planned additional changes and I believe all conversations are resolved. This is ready for final review and merge.

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019093-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019093-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

jmacato commented:
We've bikeshedded this PR so much. Next time please talk to us at the common gitter/telegram chat so that we can resolve API design before diving head first in the code.

jmacato reviewed (approved):
LGTM

jmacato committed:
Merge branch 'master' into hsvcolor

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019153-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

avaloniaui-team commented:
You can test this PR using the following package version. `0.10.999-cibuild0019153-beta`. (feed url: https://nuget.avaloniaui.net/repository/avalonia-all/index.json) [PRBUILDID]

robloo commented:
> Next time please talk to us at the common gitter/telegram chat so that we can resolve and discuss API design before diving head first in the code.

I don't plan to have these sorts of discussions in a chat room. It is better with the PR or issue. 
 1. Any parameter ordering question was not going to be noticed ahead of time. That is just something that comes later, if at all.
 2. I also created an issue for this to discuss first but there was no response.
 3. This was an existing internal class and it also follows pre-existing API. Inconsistencies in the API are inherited but were noticed until a critical review.

jmacato commented:
@robloo issues are more preferred if that's what you prefer as well. Just that we should try to avoid too much discussions on PR's since it bogs down the author's and reviewer's precious time.

jmacato commented:
in any case @MarchingCube anything we need to change here? can't merge it without your approval yet

MarchingCube reviewed (approved):
LGTM.
